apollo-configservice/src/main/java/com/ctrip/framework/apollo/configservice/service/AccessKeyServiceWithCache.java

```diff
@@ -0,0 +1,202 @@
+  private long lastTimeScanned;
+
+  private ListMultimap<String, AccessKey> accessKeyCache;
+  private ConcurrentMap<Long, AccessKey> accessKeyIdCache;
+
+  @Autowired
+  public AccessKeyServiceWithCache(AccessKeyRepository accessKeyRepository, BizConfig bizConfig) {
+    this.accessKeyRepository = accessKeyRepository;
+    this.bizConfig = bizConfig;
+
+    initialize();
+  }
+
+  private void initialize() {
+    scheduledExecutorService = new ScheduledThreadPoolExecutor(1,
+        ApolloThreadFactory.create("AccessKeyServiceWithCache", true));
+
+    ListMultimap<String, AccessKey> multimap = ListMultimapBuilder.hashKeys(128)
+        .arrayListValues().build();
+    accessKeyCache = Multimaps.synchronizedListMultimap(multimap);
+    accessKeyIdCache = Maps.newConcurrentMap();
+  }
+
+  public List<String> getAvailableSecrets(String appId) {
+    List<AccessKey> accessKeys = accessKeyCache.get(appId);
+    if (CollectionUtils.isEmpty(accessKeys)) {
+      return Collections.emptyList();
+    }
+
+    return accessKeys.stream()
+        .filter(AccessKey::isEnabled)
+        .map(AccessKey::getSecret)
+        .collect(Collectors.toList());
+  }
+
+  @Override
+  public void afterPropertiesSet() throws Exception {
+    populateDataBaseInterval();
+    scanNewAndUpdatedAccessKeys(); //block the startup process until load finished
+
+    scheduledExecutorService.scheduleWithFixedDelay(this::scanNewAndUpdatedAccessKeys,
+        scanInterval, scanInterval, scanIntervalTimeUnit);
+
+    scheduledExecutorService.scheduleAtFixedRate(this::rebuildAccessKeyCache,
+        rebuildInterval, rebuildInterval, rebuildIntervalTimeUnit);
+  }
+
+  private void scanNewAndUpdatedAccessKeys() {
+    Transaction transaction = Tracer.newTransaction("Apollo.AccessKeyServiceWithCache",
+        "scanNewAndUpdatedAccessKeys");
+    try {
+      loadNewAndUpdatedAccessKeys();
+      transaction.setStatus(Transaction.SUCCESS);
+    } catch (Throwable ex) {
+      transaction.setStatus(ex);
+      logger.error("Load new/updated app access keys failed", ex);
+    } finally {
+      transaction.complete();
+    }
+  }
+
+  private void rebuildAccessKeyCache() {
+    Transaction transaction = Tracer.newTransaction("Apollo.AccessKeyServiceWithCache",
+        "rebuildCache");
+    try {
+      deleteAccessKeyCache();
+      transaction.setStatus(Transaction.SUCCESS);
+    } catch (Throwable ex) {
+      transaction.setStatus(ex);
+      logger.error("Rebuild cache failed", ex);
+    } finally {
+      transaction.complete();
+    }
+  }
+
+  private void loadNewAndUpdatedAccessKeys() {
+    boolean hasMore = true;
+    while (hasMore && !Thread.currentThread().isInterrupted()) {
+      //current batch is 500
+      Date fromTime = new Date(lastTimeScanned);
+      List<AccessKey> accessKeys = accessKeyRepository
+          .findFirst500ByDataChangeLastModifiedTimeGreaterThanOrderByDataChangeLastModifiedTimeAsc(fromTime);
```

Comment:
I see the current logic is `where DataChange_LastTime > xxx`, but I think we need to change it to `where DataChange_LastTime >= xxx` or we may miss some records.

e.g. there are 2 records with the same DataChange_LastTime, if the last scan only contains the first record, then the next records won't be scanned forever.

But with `where DataChange_LastTime >= xxx`, it will cause the last record scanned everytime, so maybe we could add more conditions like `where DataChange_LastTime >= xxx and id != yyy`

Reply:
Thank you for your careful review, this is indeed my mistake.

But if there are a lot of records with the same `DataChange_LastTime `, maybe we should use the condition like `where DataChange_LastTime >= xxx and id not in (xxx)`.

And suppose we use `limit 500`, if there are less than 500 records with the same `DataChange_LastTime`, the condition above will work fine. If more than 1000 records with the same `DataChange_LastTime`, the condition will cause an infinite loop.

Reply:
How about I still use `where DataChange_LastTime > ? limit 500` and use `where DataChange_LastTime = ? limit 500` in each loop.

I think this is a simple way.

Reply:
You are right about the infinite loop and I think your idea to check `where DataChange_LastTime = ? in each loop if the size of the results is equal to 500` is workable.